### PR TITLE
Implemented IID Delete API

### PR DESCRIPTION
**Pull request by hiranya911**
A new API that enables deleting instance IDs associated with a Firebase project.

```
let result: Promise<void> = admin.instanceId().deleteInstanceId(iid);
```

go/firebase-admin-iid

### CODE REVIEW

File: src/instance-id/instance-id.ts [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+import {FirebaseInstanceIdRequestHandler} from './instance-id-request';
+
+import * as utils from '../utils/index';
+import * as validator from '../utils/validator';
+
+/**
+ * Internals of an InstanceId service instance.
+ */
+class InstanceIdInternals implements FirebaseServiceInternalsInterface {
+    /**
```
Comment (bojeil-google) [inline src/instance-id/instance-id.ts]: Update the indentation to be consistent. We use 2 spaces in this case.

File: src/instance-id/instance-id.ts [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+     * Deletes the service and its associated resources.
+     *
+     * @return {Promise<()>} An empty Promise that will be fulfilled when the service is deleted.
+     */
+    public delete(): Promise<void> {
+        // There are no resources to clean up
```
Comment (bojeil-google) [inline src/instance-id/instance-id.ts]: Fix indentation

File: src/instance-id/instance-id.ts [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        return Promise.resolve(undefined);
+    }
+}
+
+export class InstanceId implements FirebaseServiceInterface {
+    public INTERNAL: InstanceIdInternals = new InstanceIdInternals();
```
Comment (bojeil-google) [inline src/instance-id/instance-id.ts]: Indentation needs to be fixed here too and the rest of the class.

File: src/instance-id/instance-id.ts [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+    private app_: FirebaseApp;
```
Comment (bojeil-google) [inline src/instance-id/instance-id.ts]: Let's stick with no trailing _ for private variables. This is also the recommended style for ts.

File: src/instance-id/instance-id.ts [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    private requestHandler: FirebaseInstanceIdRequestHandler;
+
+    /**
+     * @param {Object} app The app for this Auth service.
```
Comment (bojeil-google) [inline src/instance-id/instance-id.ts]: Update comment: instance ID service?

File: src/utils/api-request.ts [hunk 6]
```diff
@@ -224,11 +224,16 @@ export class SignedApiRequestHandler extends HttpRequestHandler {
...
       httpMethod: HttpMethod,
-      data: Object,
-      headers: Object,
-      timeout: number): Promise<Object> {
+      data?: Object,
+      headers?: Object,
+      timeout?: number): Promise<Object> {
     return this.app_.INTERNAL.getToken().then((accessTokenObj) => {
-      let headersCopy: Object = deepCopy(headers);
+      let headersCopy: Object;
```
Comment (bojeil-google) [inline src/utils/api-request.ts]: Suggestion: You can also do:
`let headersCopy: Object = (headers && deepCopy(headers)) || {};`

File: test/unit/firebase-namespace.spec.ts [hunk 7]
```diff
@@ -550,4 +551,35 @@ describe('FirebaseNamespace', () => {
...
+      firebaseNamespace.initializeApp(mocks.appOptions, 'testApp');
+      expect(() => {
+        firebaseNamespace.instanceId();
+      }).to.throw(DEFAULT_APP_NOT_FOUND);
+    });
+
+    it('should return a valid namespace when the default app is initialized', () => {
+      let app: FirebaseApp = firebaseNamespace.initializeApp(mocks.appOptions);
+      let iid: InstanceId = firebaseNamespace.instanceId();
+      expect(iid).to.not.be.null;
```
Comment (bojeil-google) [inline test/unit/firebase-namespace.spec.ts]: Add check:
expect(iid.app).to.be.deep.equal(app);

File: test/unit/firebase-namespace.spec.ts [hunk 8]
```diff
@@ -550,4 +551,35 @@ describe('FirebaseNamespace', () => {
...
+    });
+
+    it('should return a valid namespace when the named app is initialized', () => {
+      let app: FirebaseApp = firebaseNamespace.initializeApp(mocks.appOptions, 'testApp');
+      let iid: InstanceId = firebaseNamespace.instanceId(app);
+      expect(iid).to.not.be.null;
```
Comment (bojeil-google) [inline test/unit/firebase-namespace.spec.ts]: Add check:
expect(iid.app).to.be.deep.equal(app);

File: test/unit/instance-id/instance-id.spec.ts [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+import {FirebaseInstanceIdError, InstanceIdClientErrorCode} from '../../../src/utils/error';
+
+import * as validator from '../../../src/utils/validator';
+
+chai.should();
+chai.use(sinonChai);
+chai.use(chaiAsPromised);
+
+describe('InstanceId', () => {
+    let iid: InstanceId;
```
Comment (bojeil-google) [inline test/unit/instance-id/instance-id.spec.ts]: fix indentation

File: test/unit/instance-id/instance-id.spec.ts [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+      const testInstanceId = 'test-iid';
+
+      afterEach(() => {
+        _.forEach(stubs, (stub) => stub.restore());
+        stubs = [];
+      });
+
+      it('should be rejected given no instance ID', () => {
+        return (iid as any).deleteInstanceId()
+        .should.eventually.be.rejected.and.have.property('code', 'instance-id/invalid-instance-id');
```
Comment (bojeil-google) [inline test/unit/instance-id/instance-id.spec.ts]: Indent second line `.should...` and `.then` below too.

File: test/unit/instance-id/instance-id.spec.ts [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+      });
+
+      it('should be rejected given an invalid instance ID', () => {
+        return iid.deleteInstanceId('')
```
Comment (bojeil-google) [inline test/unit/instance-id/instance-id.spec.ts]: Why don't you use `should.eventually.be.rejected.and.have...` like above

File: test/unit/instance-id/instance-id.spec.ts [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+        stubs.push(stub);
+        return iid.deleteInstanceId(testInstanceId)
+          .then((result) => {
+            // Confirm underlying API called with expected parameters.
+            expect(stub).to.have.been.calledOnce.and.calledWith(testInstanceId);
+          });
+      });
+
+      it('should throw an error when the backend returns an error', () => {
+        // Stub getAccountInfoByUid to throw a backend error.
```
Comment (bojeil-google) [inline test/unit/instance-id/instance-id.spec.ts]: Update comment `getAccountInfoByUid`

File: src/instance-id/instance-id-request.ts [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+const FIREBASE_IID_HOST = 'console.firebase.google.com';
+/** Firebase IID backend port number. */
+const FIREBASE_IID_PORT = 443;
+/** Firebase IID backend path. */
+const FIREBASE_IID_PATH = '/v1/';
+/** Firebase IID request timeout duration in milliseconds. */
+const FIREBASE_IID_TIMEOUT = 10000;
+
+/**
+ * Class that provides mechanism to send requests to the Firebase Auth backend endpoints.
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/instance-id/instance-id-request.ts]: Done

File: src/instance-id/instance-id.ts (hunk 1 above)
Comment (hiranya911) [reply src/instance-id/instance-id.ts]: Done

File: src/instance-id/instance-id.ts (hunk 2 above)
Comment (hiranya911) [reply src/instance-id/instance-id.ts]: Done

File: src/instance-id/instance-id.ts (hunk 3 above)
Comment (hiranya911) [reply src/instance-id/instance-id.ts]: Done

File: src/instance-id/instance-id.ts (hunk 4 above)
Comment (hiranya911) [reply src/instance-id/instance-id.ts]: That is going to conflict with the `get app()` getter. Is it better to remove that, and expose `app` as a public attribute of this class?

File: src/instance-id/instance-id.ts (hunk 5 above)
Comment (hiranya911) [reply src/instance-id/instance-id.ts]: Done

File: src/utils/api-request.ts (hunk 6 above)
Comment (hiranya911) [reply src/utils/api-request.ts]: Much cleaner. Thanks.

File: test/unit/instance-id/instance-id.spec.ts (hunk 9 above)
Comment (hiranya911) [reply test/unit/instance-id/instance-id.spec.ts]: Done

File: test/unit/instance-id/instance-id.spec.ts (hunk 10 above)
Comment (hiranya911) [reply test/unit/instance-id/instance-id.spec.ts]: Done

File: test/unit/instance-id/instance-id.spec.ts (hunk 12 above)
Comment (hiranya911) [reply test/unit/instance-id/instance-id.spec.ts]: Done

File: test/unit/instance-id/instance-id.spec.ts (hunk 11 above)
Comment (hiranya911) [reply test/unit/instance-id/instance-id.spec.ts]: Done

File: test/unit/firebase-namespace.spec.ts (hunk 8 above)
Comment (hiranya911) [reply test/unit/firebase-namespace.spec.ts]: Done

File: test/unit/firebase-namespace.spec.ts (hunk 7 above)
Comment (hiranya911) [reply test/unit/firebase-namespace.spec.ts]: Done

**Comment by hiranya911** [comment]
Made the suggested changes, and left one question. Over to @bojeil-google  for another look.

**Comment by hiranya911** [comment]
@bojeil-google I pushed a couple of commits to improve error handling and test coverage. PTAL.

File: test/unit/instance-id/instance-id-request.spec.ts [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+
+  describe('Constructor', () => {
+    it('should succeed with a FirebaseApp instance', () => {
+      expect(() => {
+        return new FirebaseInstanceIdRequestHandler(mockApp, projectId);
+      }).not.to.throw(Error);
+    });
+  });
+
+  describe('deleteInstanceId', () => {
```
Comment (bojeil-google) [inline test/unit/instance-id/instance-id-request.spec.ts]: Looks good. One suggestion: i think it would be a good idea to also confirm the error code too.

Comment (hiranya911) [reply test/unit/instance-id/instance-id-request.spec.ts]: Done

**Pull request merged**